File: kotlin/src/main/kotlin/com/tokenpack/TokenPack.kt
package com.tokenpack

import com.google.gson.*
import kotlin.math.floor
import kotlin.math.max

/**
 * TokenPack - Pack JSON data into token-efficient CSV for LLM prompts.
 *
 * ```
 * val csv = TokenPack.pack(jsonString)        // JSON -> CSV (fewer tokens)
 * val arr = TokenPack.unpack(csv)             // CSV -> JSON (back to original)
 * val prompt = TokenPack.packForPrompt("Analyze:", jsonString)
 * ```
 */
object TokenPack {

    private val gson = Gson()

    // -------------------------------------------------------------------
    // Public API
    // -------------------------------------------------------------------

    /**
     * Convert a JSON string to the most token-efficient text format.
     * Arrays of similar objects -> CSV. Everything else -> compact JSON.
     */
    fun pack(json: String): String {
        val el = try {
            JsonParser.parseString(json)
        } catch (_: JsonSyntaxException) {
            return json
        }
        return packElement(el)
    }

    /**
     * Convert a JsonElement to the most token-efficient text format.
     */
    fun pack(element: JsonElement): String = packElement(element)

    /**
     * Convert packed text (CSV or JSON) back to a JsonArray.
     */
    fun unpack(text: String): JsonArray {
        val trimmed = text.trim()
        if (trimmed.isEmpty()) return JsonArray()

        return when (trimmed[0]) {
            '[' -> JsonParser.parseString(trimmed).asJsonArray
            '{' -> JsonArray().apply {
                add(JsonParser.parseString(trimmed).asJsonObject)
            }
            else -> fromCsv(trimmed)
        }
    }

    /**
     * Combine a message with packed data.
     */
    fun packForPrompt(message: String, json: String): String =
        "$message\n${pack(json)}"

    fun packForPrompt(message: String, element: JsonElement): String =
        "$message\n${pack(element)}"

    // -------------------------------------------------------------------
    // Internal: shape detection
    // -------------------------------------------------------------------

    private fun packElement(el: JsonElement): String {
        if (!el.isJsonArray) return gson.toJson(el)

        val arr = el.asJsonArray
        if (arr.size() < 2) return gson.toJson(el)
        if (!isPackableArray(arr)) return gson.toJson(el)

        return toCsv(arr)
    }

    private fun isPackableArray(arr: JsonArray): Boolean {
        val allKeys = linkedSetOf<String>()
        val rowKeys = mutableListOf<Set<String>>()

        for (item in arr) {
            if (!item.isJsonObject) return false
            val keys = item.asJsonObject.keySet()
            allKeys.addAll(keys)
            rowKeys.add(keys)
        }

        if (allKeys.isEmpty()) return false

        // Shared keys = intersection of all row key sets
        val shared = linkedSetOf<String>().apply { addAll(rowKeys[0]) }
        for (rk in rowKeys) {
            shared.retainAll(rk)
        }
        if (shared.isEmpty()) return false

        val threshold = max(allKeys.size * 0.3, 1.0)
        return rowKeys.all { it.size >= threshold }
    }

    // -------------------------------------------------------------------
    // Internal: JSON -> CSV
    // -------------------------------------------------------------------

    private fun toCsv(arr: JsonArray): String {
        val flatRows = arr.map { flattenRow(it.asJsonObject) }

        // Ordered superset of headers
        val headerSet = linkedSetOf<String>()
        for (row in flatRows) {
            headerSet.addAll(row.keys)
        }
        val headers = headerSet.toList()

        return buildString {
            append(csvLine(headers))
            for (row in flatRows) {
                append('\n')
                val values = headers.map { h -> row.getOrDefault(h, "") }
                append(csvLine(values))
            }
        }
    }

    private fun flattenRow(obj: JsonObject): Map<String, String> {
        val flat = linkedMapOf<String, String>()
        flattenValue(obj, "", flat)
        return flat
    }

    private fun flattenValue(value: JsonElement, prefix: String, out: MutableMap<String, String>) {
        when {
            value.isJsonObject -> {
                for ((key, child) in value.asJsonObject.entrySet()) {
                    val fullKey = if (prefix.isEmpty()) key else "$prefix.$key"
                    flattenValue(child, fullKey, out)
                }
            }
            value.isJsonArray -> {
                val arr = value.asJsonArray
                val allPrimitive = arr.all { !it.isJsonObject && !it.isJsonArray }
                if (allPrimitive) {
                    out[prefix] = arr.joinToString("|") { formatCell(it) }
                } else {
                    out[prefix] = gson.toJson(arr)
                }
            }
            else -> {
                out[prefix] = formatCell(value)
            }
        }
    }

    private fun formatCell(el: JsonElement?): String {
        if (el == null || el.isJsonNull) return ""
        if (el.isJsonPrimitive) {
            val p = el.asJsonPrimitive
            return when {
                p.isBoolean -> if (p.asBoolean) "true" else "false"
                p.isNumber -> {
                    val d = p.asDouble
                    if (d == floor(d) && !d.isInfinite()) {
                        d.toLong().toString()
                    } else {
                        p.asString
                    }
                }
                else -> p.asString
            }
        }
        return gson.toJson(el)
    }

    private fun csvLine(fields: List<String>): String =
        fields.joinToString(",") { csvEscape(it) }

    private fun csvEscape(value: String): String =
        if (value.contains(",") || value.contains("\"") || value.contains("\n") || value.contains("\r")) {
            "\"${value.replace("\"", "\"\"")}\""
        } else {
            value
        }

    // -------------------------------------------------------------------
    // Internal: CSV -> JSON (unpack)
    // -------------------------------------------------------------------

    private fun fromCsv(text: String): JsonArray {
        val lines = parseCsvLines(text)
        if (lines.size < 2) return JsonArray()

        val headers = lines[0]
        val result = JsonArray()

        for (i in 1 until lines.size) {
            val row = lines[i]
            if (row.all { it.isEmpty() }) continue

            val flat = linkedMapOf<String, JsonElement>()
            for (j in headers.indices) {
                val raw = if (j < row.size) row[j] else ""
                flat[headers[j]] = parseCell(raw)
            }
            result.add(unflatten(flat))
        }

        return result
    }

    private fun parseCell(raw: String): JsonElement {
        if (raw.isEmpty()) return JsonNull.INSTANCE

        when (raw) {
            "true" -> return JsonPrimitive(true)
            "false" -> return JsonPrimitive(false)
        }

        // Try integer
        if (raw.matches(Regex("-?\\d+"))) {
            raw.toLongOrNull()?.let { return JsonPrimitive(it) }
        }

        // Try float
        if (raw.matches(Regex("-?\\d+\\.\\d+"))) {
            raw.toDoubleOrNull()?.let { return JsonPrimitive(it) }
        }

        // Pipe-separated array
        if ('|' in raw && !raw.startsWith("[")) {
            return JsonArray().apply {
                for (part in raw.split("|", limit = -1)) {
                    add(parseCell(part))
                }
            }
        }

        // Embedded JSON
        if (raw.startsWith("[") || raw.startsWith("{")) {
            try {
                return JsonParser.parseString(raw)
            } catch (_: JsonSyntaxException) {
                // fall through
            }
        }

        return JsonPrimitive(raw)
    }

    private fun unflatten(flat: Map<String, JsonElement>): JsonObject {
        val result = JsonObject()
        for ((key, value) in flat) {
            val parts = key.split(".")
            var current = result
            for (i in 0 until parts.size - 1) {
                if (!current.has(parts[i]) || !current[parts[i]].isJsonObject) {
                    current.add(parts[i], JsonObject())
                }
                current = current.getAsJsonObject(parts[i])
            }
            current.add(parts.last(), value)
        }
        return result
    }

    private fun parseCsvLines(text: String): List<List<String>> {
        val lines = mutableListOf<List<String>>()
        var current = mutableListOf<String>()
        val field = StringBuilder()
        var inQuotes = false

        var i = 0
        while (i < text.length) {
            val ch = text[i]

            if (inQuotes) {
                if (ch == '"') {
                    if (i + 1 < text.length && text[i + 1] == '"') {
                        field.append('"')
                        i++
                    } else {
                        inQuotes = false
                    }
                } else {
                    field.append(ch)
                }
            } else {
                when (ch) {
                    '"' -> inQuotes = true
                    ',' -> {
                        current.add(field.toString())
                        field.setLength(0)
                    }
                    '\n' -> {
                        current.add(field.toString())
                        field.setLength(0)
                        lines.add(current)
                        current = mutableListOf()
                    }
                    '\r' -> { /* skip */ }
                    else -> field.append(ch)
                }
            }
            i++
        }

        current.add(field.toString())
        if (current.any { it.isNotEmpty() }) {
            lines.add(current)
        }

        return lines
    }
}

File: swift/Sources/TokenPack/TokenPack.swift
/// TokenPack - Pack JSON data into token-efficient CSV for LLM prompts.
///
/// Usage:
///   let csv = TokenPack.pack(myData)        // JSON → CSV (fewer tokens)
///   let json = TokenPack.unpack(csv)         // CSV → back to dicts
///   let prompt = TokenPack.packForPrompt("Analyze:", data: myData)

import Foundation

// MARK: - Public API

public enum TokenPack {

    /// Convert JSON-compatible data to the most token-efficient text format.
    ///
    /// - Parameter data: `[[String: Any]]`, a JSON string, or any JSON-serializable value.
    /// - Returns: CSV string for packable arrays, compact JSON otherwise.
    public static func pack(_ data: Any) -> String {
        var resolved = data

        // If data is a String, try parsing as JSON first
        if let str = data as? String {
            guard let jsonData = str.data(using: .utf8),
                  let parsed = try? JSONSerialization.jsonObject(with: jsonData, options: []) else {
                return str
            }
            resolved = parsed
        }

        // If array of dicts with >= 2 items and packable -> CSV
        if let array = resolved as? [[String: Any]], array.count >= 2, isPackableArray(array) {
            return toCsv(array)
        }

        // Fallback: compact JSON
        if let jsonData = try? JSONSerialization.data(withJSONObject: resolved, options: [.sortedKeys, .fragmentsAllowed]),
           let jsonStr = String(data: jsonData, encoding: .utf8) {
            return jsonStr
        }

        return "\(resolved)"
    }

    /// Convert packed text back to JSON-compatible objects.
    ///
    /// - Parameter text: CSV (produced by `pack`) or JSON string.
    /// - Returns: Array of dictionaries.
    public static func unpack(_ text: String) -> [[String: Any]] {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return [] }

        let first = trimmed.first!
        if first == "[" || first == "{" {
            if let data = trimmed.data(using: .utf8),
               let parsed = try? JSONSerialization.jsonObject(with: data, options: []) {
                if let array = parsed as? [[String: Any]] {
                    return array
                }
                // Single dict -> wrap in array
                if let dict = parsed as? [String: Any] {
                    return [dict]
                }
            }
            return []
        }

        // Otherwise treat as CSV
        return fromCsv(trimmed)
    }

    /// Convenience: combine a user message with packed data.
    public static func packForPrompt(_ message: String, data: Any) -> String {
        let packed = pack(data)
        return "\(message)\n\(packed)"
    }
}

// MARK: - Shape Detection

private func isPackableArray(_ data: [[String: Any]]) -> Bool {
    if data.isEmpty { return false }

    // Collect all keys across all rows
    var allKeys = Set<String>()
    for item in data {
        for key in item.keys {
            allKeys.insert(key)
        }
    }
    if allKeys.isEmpty { return false }

    // Shared keys: intersection of all rows' key sets
    var sharedKeys = Set(data[0].keys)
    for item in data {
        sharedKeys.formIntersection(item.keys)
    }
    if sharedKeys.isEmpty { return false }

    // Each row must have at least 30% of the superset key count
    let threshold = max(Double(allKeys.count) * 0.3, 1.0)
    for item in data {
        if Double(item.count) < threshold {
            return false
        }
    }

    return true
}

// MARK: - CSV Writing

private func flattenValue(_ value: Any, prefix: String) -> [(String, Any?)] {
    // Nested dict -> dot notation
    if let dict = value as? [String: Any] {
        var out: [(String, Any?)] = []
        for (k, v) in dict.sorted(by: { $0.key < $1.key }) {
            let fullKey = prefix.isEmpty ? k : "\(prefix).\(k)"
            out.append(contentsOf: flattenValue(v, prefix: fullKey))
        }
        return out
    }

    // Array
    if let array = value as? [Any] {
        // All primitives -> pipe-join
        if array.allSatisfy({ isPrimitive($0) }) {
            let joined = array.map { primToStr($0) }.joined(separator: "|")
            return [(prefix, joined)]
        }
        // Complex array -> JSON fallback
        if let jsonData = try? JSONSerialization.data(withJSONObject: array, options: [.sortedKeys]),
           let jsonStr = String(data: jsonData, encoding: .utf8) {
            return [(prefix, jsonStr)]
        }
        return [(prefix, nil)]
    }

    return [(prefix, value)]
}

private func flattenRow(_ row: [String: Any]) -> [(String, Any?)] {
    var flat: [(String, Any?)] = []
    for (key, value) in row.sorted(by: { $0.key < $1.key }) {
        flat.append(contentsOf: flattenValue(value, prefix: key))
    }
    return flat
}

private func toCsv(_ data: [[String: Any]]) -> String {
    // Flatten all rows
    let flatRows: [[(String, Any?)]] = data.map { flattenRow($0) }

    // Collect ordered superset of keys (preserving first-seen order)
    var seen = [String]()
    var seenSet = Set<String>()
    for row in flatRows {
        for (key, _) in row {
            if !seenSet.contains(key) {
                seenSet.insert(key)
                seen.append(key)
            }
        }
    }
    let headers = seen

    // Build lookup dicts for each row
    let rowDicts: [[String: Any?]] = flatRows.map { pairs in
        var dict = [String: Any?]()
        for (k, v) in pairs {
            dict[k] = v
        }
        return dict
    }

    var lines: [String] = []
    lines.append(csvLine(headers))
    for row in rowDicts {
        let cells = headers.map { h -> String in
            if let val = row[h] {
                return formatCell(val)
            }
            return ""
        }
        lines.append(csvLine(cells))
    }

    return lines.joined(separator: "\n")
}

private func formatCell(_ value: Any?) -> String {
    guard let value = value else { return "" }

    if value is NSNull { return "" }

    if isBool(value), let b = value as? Bool { return b ? "true" : "false" }
    if let n = value as? Int { return "\(n)" }
    if let n = value as? Double {
        // If it's a whole number, format without decimal
        if n == n.rounded() && !n.isInfinite && !n.isNaN && abs(n) < 1e15 {
            return "\(Int(n))"
        }
        return "\(n)"
    }
    if let s = value as? String { return s }

    // Fallback: JSON
    if let jsonData = try? JSONSerialization.data(withJSONObject: value, options: [.sortedKeys]),
       let jsonStr = String(data: jsonData, encoding: .utf8) {
        return jsonStr
    }
    return "\(value)"
}

private func csvLine(_ fields: [String]) -> String {
    return fields.map { csvEscape($0) }.joined(separator: ",")
}

private func csvEscape(_ value: String) -> String {
    if value.contains(",") || value.contains("\"") || value.contains("\n") || value.contains("\r") {
        let escaped = value.replacingOccurrences(of: "\"", with: "\"\"")
        return "\"\(escaped)\""
    }
    return value
}

// MARK: - CSV Parsing (unpack)

private func fromCsv(_ text: String) -> [[String: Any]] {
    let lines = parseCsvLines(text)
    if lines.count < 2 { return [] }

    let headers = lines[0]
    var rows: [[String: Any]] = []

    for i in 1..<lines.count {
        let csvRow = lines[i]
        if csvRow.allSatisfy({ $0.isEmpty }) { continue }

        var flat: [String: Any] = [:]
        for (j, header) in headers.enumerated() {
            let raw = j < csvRow.count ? csvRow[j] : ""
            if let parsed = parseCell(raw) {
                flat[header] = parsed
            }
        }
        rows.append(unflatten(flat))
    }

    return rows
}

private func parseCell(_ raw: String) -> Any? {
    // Empty -> nil
    if raw.isEmpty { return nil }

    // Booleans
    if raw == "true" { return true }
    if raw == "false" { return false }

    // Integer
    if let intVal = Int(raw) {
        // Make sure the string representation matches (no leading zeros etc.)
        if "\(intVal)" == raw {
            return intVal
        }
    }

    // Double
    if raw.contains("."), let dblVal = Double(raw) {
        return dblVal
    }

    // Pipe-separated array
    if raw.contains("|") && !raw.hasPrefix("[") {
        let parts = raw.components(separatedBy: "|")
        return parts.map { parseCell($0) as Any }
    }

    // Embedded JSON array/object
    if raw.hasPrefix("[") || raw.hasPrefix("{") {
        if let data = raw.data(using: .utf8),
           let parsed = try? JSONSerialization.jsonObject(with: data, options: []) {
            return parsed
        }
    }

    // Plain string
    return raw
}

private func unflatten(_ flat: [String: Any]) -> [String: Any] {
    var result: [String: Any] = [:]
    for (key, value) in flat {
        let parts = key.components(separatedBy: ".")
        setNestedValue(&result, parts: parts, index: 0, value: value)
    }
    return result
}

/// Recursively set a value in a nested dictionary using key path parts.
private func setNestedValue(_ dict: inout [String: Any], parts: [String], index: Int, value: Any) {
    let key = parts[index]
    if index == parts.count - 1 {
        // Leaf: set the value directly
        dict[key] = value
    } else {
        // Intermediate: ensure a nested dict exists and recurse
        var nested = dict[key] as? [String: Any] ?? [:]
        setNestedValue(&nested, parts: parts, index: index + 1, value: value)
        dict[key] = nested
    }
}

private func parseCsvLines(_ text: String) -> [[String]] {
    var lines: [[String]] = []
    var current: [String] = []
    var field = ""
    var inQuotes = false

    let chars = Array(text)
    var i = 0

    while i < chars.count {
        let ch = chars[i]

        if inQuotes {
            if ch == "\"" {
                if i + 1 < chars.count && chars[i + 1] == "\"" {
                    field.append("\"")
                    i += 2
                    continue
                } else {
                    inQuotes = false
                    i += 1
                    continue
                }
            } else {
                field.append(ch)
            }
        } else {
            if ch == "\"" {
                inQuotes = true
            } else if ch == "," {
                current.append(field)
                field = ""
            } else if ch == "\n" {
                current.append(field)
                field = ""
                lines.append(current)
                current = []
            } else if ch == "\r" {
                // skip carriage return
            } else {
                field.append(ch)
            }
        }

        i += 1
    }

    // Flush last field and line
    current.append(field)
    if current.contains(where: { !$0.isEmpty }) {
        lines.append(current)
    }

    return lines
}

// MARK: - Helpers

/// Distinguish true Bool from NSNumber-bridged integers.
/// In Foundation, Bool bridges to NSNumber and `as? Bool` can match integers.
/// CFBooleanGetTypeID lets us detect actual booleans.
private func isBool(_ value: Any) -> Bool {
    guard let number = value as? NSNumber else { return false }
    return CFGetTypeID(number) == CFBooleanGetTypeID()
}

private func isPrimitive(_ value: Any) -> Bool {
    if value is NSNull { return true }
    if value is String { return true }
    if isBool(value) { return true }
    if value is Int { return true }
    if value is Double { return true }
    // NSNumber that isn't Bool
    if value is NSNumber { return true }
    return false
}

private func primToStr(_ value: Any) -> String {
    if value is NSNull { return "" }
    if isBool(value), let b = value as? Bool { return b ? "true" : "false" }
    if let s = value as? String { return s }
    if let n = value as? Int { return "\(n)" }
    if let n = value as? Double { return "\(n)" }
    return ""
}
